packages/core/src/types.ts

```diff
@@ -409,7 +409,7 @@ export type StateNodesConfig<
 > = {
   [K in keyof TStateSchema['states']]: StateNode<
     TContext,
-    TStateSchema['states'][K],
+    TStateSchema['states'][K] & {},
```

Comment:
Those `& {}` here ensure that TS understands that this type argument will be at least non-nullable (I didn't use `NonNullable` because in older versions of TS it was a conditional type and it could complicate assignability rules there). It also makes the whole intersection have `{}` or `StateSchema` constraint and both are assignable to `StateSchema` that is expected on the target side here (the required constraint within `StateNode`).

Perhaps a more clean solution would be to use `NonNullable` around `TStateSchema['states']` but that doesn't currently work in TS. I have prepared a bug fix for it.

However, all of this is related to `TStateSchema` so we won't even need that bug fix to land - we can go with this workaround here, n v4. In v5 we have already removed `TStateSchema` so this will just be removed when merging this to that branch